golang/cosmos/x/swingset/keeper/store_snapshotter.go

```diff
@@ -0,0 +1,601 @@
+	// A flag indicating whether "export data" should be part of the SwingStore export
+	// If false, the resulting SwingStoreExportProvider's GetExportData will
+	// error with io.EOF
+	IncludeExportData bool `json:"includeExportData,omitempty"`
+}
+
+// SwingStoreImportOptions are configurable options provided to the JS SwingStore import
+type SwingStoreImportOptions struct {
+	// A flag indicating whether the SwingStore import should attempt to load
+	// all historical artifacts available from the export provider
+	IncludeHistorical bool `json:"includeHistorical,omitempty"`
+}
+
+type restoreOptions struct {
+	ExportDir         string `json:"exportDir"`
+	IncludeHistorical bool   `json:"includeHistorical,omitempty"`
+}
+
+// SwingStoreSnapshotter manages the exclusive communication with the JS side
+// for SwingStore snapshots, ensuring insensitivity to sub-block timing, and
+// enforcing concurrency requirements.
+// The caller of this submodule must arrange block level commit synchronization,
+// to ensure the results are deterministic.
+//
+// Some `blockingSend` calls performed by this submodule are non-deterministic.
+// This submodule will send messages to JS from goroutines at unpredictable
+// times, but this is safe because when handling the messages, the JS side
+// does not perform operations affecting consensus and ignores state changes
+// since committing the previous block.
+// Some other `blockingSend` calls however do change the JS swing-store and
+// must happen before the Swingset controller on the JS side was inited.
+type SwingStoreSnapshotter struct {
+	logger       log.Logger
+	blockingSend func(action vm.Jsonable, mustNotBeInited bool) (string, error)
+	// Only modified by the main goroutine.
+	activeSnapshot *activeSnapshot
+}
+
+// NewSwingStoreSnapshotter creates a new SwingStoreSnapshotter
+func NewSwingStoreSnapshotter(logger log.Logger, blockingSend func(action vm.Jsonable, mustNotBeInited bool) (string, error)) SwingStoreSnapshotter {
+	return SwingStoreSnapshotter{
+		logger:         logger.With("module", fmt.Sprintf("x/%s", types.ModuleName), "submodule", "SwingStoreSnapshotter"),
+		blockingSend:   blockingSend,
+		activeSnapshot: nil,
+	}
+}
+
+// checkNotActive returns an error if there is an active snapshot.
+func (snapshotter *SwingStoreSnapshotter) checkNotActive() error {
+	active := snapshotter.activeSnapshot
+	if active != nil {
+		select {
+		case <-active.done:
+			snapshotter.activeSnapshot = nil
+		default:
+			if active.isRestore {
+				return fmt.Errorf("snapshot restore already in progress for height %d", active.height)
+			} else {
+				return fmt.Errorf("snapshot already in progress for height %d", active.height)
+			}
+		}
+	}
+	return nil
+}
+
+// InitiateSnapshot initiates a SwingStore snapshot for the given
+// height. If a snapshot is already in progress, this will fail. The snapshot
+// processing is delegated to the provided `snapshotTaker`.
```

Comment:
```suggestion
// InitiateSnapshot synchronously verifies that there is not already a snapshot
// in progress and records the start of a new one for the provided block height,
// then launches a goroutine to asynchronously coordinate with JS for initiating
// and retrieving a snapshot, sending it to the provided `snapshotTaker` for processing,
// and ultimately releasing it.
```

That the behavior includes retrieval and processing also suggests to me that "initiate" is not the best name, although a similar argument would probably apply to using the name "take" for a function that synchronously starts a process which finishes asynchronously. It might be best to just reify the separation such that e.g. call sites look like
```go
if err := snapshotter.InitiateSnapshot(height); err != nil {
	…
}
go snapshotter.FinishSnapshot(taker, exportOptions)
```

Reply:
That shape is not possible. Let's chat